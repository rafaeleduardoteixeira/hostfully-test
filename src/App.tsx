import React from 'react';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import createdRoutes from './constants/router';
import { Header } from './components/Header/Header';
import { AppContainer } from './App.styles';

function App() {
  return (
    <AppContainer>
      <Header />
      <RouterProvider router={createdRoutes} />
    </AppContainer>
  );
}

export default App;
