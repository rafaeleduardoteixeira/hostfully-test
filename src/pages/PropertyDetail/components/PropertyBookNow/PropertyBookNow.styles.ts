import styled, { css } from 'styled-components';
import { colors, media } from '../../../../themes';

export const PropertyBookNowContainer = styled.div`
  width: 30%;
  height: 400px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;

  border-radius: 10px;
  border: 1px solid rgb(221, 221, 221);
  box-shadow: rgba(0, 0, 0, 0.12) 0px 6px 16px;

  @media ${media.tablet} {
    width: 40%;
  }

  @media ${media.mobile} {
    width: 100%;
  }
`;

export const PropertyBookNowPrice = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${colors.hostfullyPrimaryColor};
  font-size: 16px;
  font-weight: 700;
  width: 100%;
  color: white;
  font-size: 2rem;
  height: 55px;
  border-top-right-radius: 10px;
  border-top-left-radius: 10px;
`;

export const PropertyBookNowButton = styled.button<{ disabled: boolean }>`
  background-color: ${colors.hostfullyPrimaryColor};
  border: none;
  padding: 20px;
  cursor: pointer;
  font-size: 16px;
  font-weight: 700;
  color: white;
  width: 100%;
  height: 55px;
  border-bottom-right-radius: 10px;
  border-bottom-left-radius: 10px;
  transition: background-color 0.3s;

  &:hover {
    background-color: ${colors.hostfullySecondaryColor};
  }

  ${(props) =>
    props.disabled &&
    css`
      background-color: ${colors.gray};
      cursor: not-allowed;

      &:hover {
        background-color: ${colors.gray};
      }
    `}
`;

export const PropertyBookNowFields = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 90%;
  gap: 15px;

  @media ${media.notebook} {
    input {
      width: 250px;
    }
    .date-picker {
      width: 240px;
    }
  }

  @media ${media.tablet} {
    input {
      width: 300px;
    }
    .date-picker {
      width: 290px;
    }
  }

  @media ${media.mobile} {
    > div {
      width: 260px;
    }
    input {
      width: 250px;
    }
    .date-picker {
      width: 240px;
    }
  }
`;

export const PropertyBookNowSuccessContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 15px;
`;

export const PropertyBookNowSuccessMessage = styled.p`
  font-size: 16px;
  font-weight: 700;
  color: ${colors.grayDark};
`;

export const PropertyBookNowSuccessButton = styled.button`
  background-color: ${colors.hostfullyPrimaryColor};
  border: none;
  padding: 10px 20px;
  cursor: pointer;
  font-size: 16px;
  font-weight: 700;
  color: white;
  border-radius: 10px;
  transition: background-color 0.3s;

  &:hover {
    background-color: ${colors.hostfullySecondaryColor};
  }
`;
