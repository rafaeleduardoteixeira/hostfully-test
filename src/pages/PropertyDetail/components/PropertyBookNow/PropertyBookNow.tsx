import React, { useContext, useState } from 'react';
import { Property } from '../../../../interfaces/property';
import {
  PropertyBookNowButton,
  PropertyBookNowContainer,
  PropertyBookNowPrice,
  PropertyBookNowFields,
  PropertyBookNowSuccessContainer,
  PropertyBookNowSuccessMessage,
  PropertyBookNowSuccessButton,
} from './PropertyBookNow.styles';
import { Input } from '../../../../components/Input/Input';
import { AppContext, AppContextType } from '../../../../Context/App.context';
import { Skeleton } from '../../../../components/Skeleton/Skeleton';
import { useNavigate } from 'react-router-dom';
import { v4 as uuidv4 } from 'uuid';
import { InputDateRange } from '../../../../components/InputDateRange/InputDateRange';
import moment from 'moment';

interface PropertyBookNowProps {
  property: Property;
}

type FormDataType = {
  name: string;
  email: string;
  checkIn: Date | null;
  checkOut: Date | null;
};

export const PropertyBookNow = ({ property }: PropertyBookNowProps) => {
  const navigate = useNavigate();
  const { isLoading, handleAddRentals, getPropertyRentedDates } = useContext(AppContext) as AppContextType;
  const [isSuccess, setIsSuccess] = useState<boolean>(false);

  const rentedDates = getPropertyRentedDates(property.id);

  const [formData, setFormData] = useState<FormDataType>({
    name: '',
    email: '',
    checkIn: null,
    checkOut: null,
  });

  const [formErrors, setFormErrors] = useState({
    name: '',
    email: '',
    checkIn: '',
    checkOut: '',
  });

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFormData({
      ...formData,
      [event.target.name]: event.target.value,
    });
  };

  const handleCheckRentedDates = (checkIn: Date, checkOut: Date) => {
    // Create to prevent that user select a date that is already rented. For instance if the user select 23/07/2024 to 27/07/2024 and the property is already rented from 24/07/2024 to 26/07/2024, the user can't select this date. I set the checkIn date the value of checkOut date.
    if (checkIn && checkOut) {
      return rentedDates?.some((date) => {
        if (date.start && date.end) {
          if (
            (checkIn >= date.start && checkIn <= date.end) ||
            (checkOut >= date.start && checkOut <= date.end) ||
            (checkIn <= date.start && checkOut >= date.end)
          )
            return true;
        }
        return false;
      });
    }
    return false;
  };

  const handleChangeDate = (value: Date[]) => {
    if (handleCheckRentedDates(value[0], value[1])) {
      setFormData({
        ...formData,
        checkIn: null,
        checkOut: null,
      });
      return;
    }
    if (value) {
      setFormData({
        ...formData,
        checkIn: value[0],
        checkOut: value[1],
      });
      return;
    }
    setFormData({
      ...formData,
      checkIn: null,
      checkOut: null,
    });
  };

  const validateForm = () => {
    if (!formData.name || !formData.email || !formData.checkIn || !formData.checkOut) {
      setFormErrors({
        ...formErrors,
        name: !formData.name ? 'Name is required' : '',
        email: !formData.email ? 'Email is required' : '',
        checkIn: !formData.checkIn ? 'Invalid date' : '',
        checkOut: !formData.checkOut ? 'Invalid date' : '',
      });
      return false;
    }
    return true;
  };

  const handleAddRental = async () => {
    if (!validateForm()) return;
    setIsSuccess(
      await handleAddRentals({
        ...formData,
        id: uuidv4(),
        propertyId: property.id,
      }),
    );
  };

  const handleSeeMyTrips = () => {
    navigate('/my-trips');
  };

  return (
    <PropertyBookNowContainer data-testid="PropertyBookNow">
      <PropertyBookNowPrice>${property.price}</PropertyBookNowPrice>
      <PropertyBookNowFields>
        {isLoading ? (
          <>
            <Skeleton width="300px" height="50px" padding="0 10px 0 20px" />
            <Skeleton width="300px" height="50px" padding="0 10px 0 20px" />
            <Skeleton width="300px" height="50px" padding="0 10px 0 20px" />
          </>
        ) : isSuccess ? (
          <PropertyBookNowSuccessContainer>
            <PropertyBookNowSuccessMessage>Rental added successfully!</PropertyBookNowSuccessMessage>
            <PropertyBookNowSuccessButton onClick={handleSeeMyTrips}>See my trips!</PropertyBookNowSuccessButton>
          </PropertyBookNowSuccessContainer>
        ) : (
          <>
            <Input
              name="name"
              placeholder="Name"
              type="text"
              onChange={handleChange}
              width="300px"
              error={formErrors.name}
            />
            <Input
              name="email"
              placeholder="Email"
              type="email"
              onChange={handleChange}
              width="300px"
              error={formErrors.email}
            />
            <InputDateRange
              bookedDates={rentedDates}
              dateStart={formData.checkIn ? moment(formData.checkIn, 'America/Los Angeles').toDate() : undefined}
              dateEnd={formData.checkOut ? moment(formData.checkOut, 'America/Los Angeles').toDate() : undefined}
              onChange={handleChangeDate}
              error={formErrors.checkIn || formErrors.checkOut}
            />
          </>
        )}
      </PropertyBookNowFields>
      <PropertyBookNowButton onClick={handleAddRental} disabled={isSuccess || isLoading}>
        Book now
      </PropertyBookNowButton>
    </PropertyBookNowContainer>
  );
};
