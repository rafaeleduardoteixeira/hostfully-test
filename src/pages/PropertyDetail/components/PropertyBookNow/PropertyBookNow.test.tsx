import React, { act } from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { PropertyBookNow } from './PropertyBookNow';
import { AppContext } from '../../../../Context/App.context';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

jest.mock('react-leaflet', () => jest.fn());

describe('PropertyBookNow', () => {
  test('Snapshot test', () => {
    const { asFragment } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyBookNow
          property={{
            id: '1',
            name: 'Name 1',
            description: 'Description 1',
            city: 'City 1',
            country: 'Country 1',
            purpose: 'Purpose 1',
            price: 1,
            images: ['image'],
            position: {
              latitude: 0,
              longitude: 0,
            },
          }}
        />
      </AppContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Should render PropertyBookNow', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyBookNow
          property={{
            id: '1',
            name: 'Name 1',
            description: 'Description 1',
            city: 'City 1',
            country: 'Country 1',
            purpose: 'Purpose 1',
            price: 1,
            images: ['image'],
            position: {
              latitude: 0,
              longitude: 0,
            },
          }}
        />
      </AppContext.Provider>,
    );

    const PropertyBookNowElement = screen.getByTestId('PropertyBookNow');
    expect(PropertyBookNowElement).toBeInTheDocument();
  });
  test('Should  check is loading', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: true,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyBookNow
          property={{
            id: '1',
            name: 'Name 1',
            description: 'Description 1',
            city: 'City 1',
            country: 'Country 1',
            purpose: 'Purpose 1',
            price: 1,
            images: ['image'],
            position: {
              latitude: 0,
              longitude: 0,
            },
          }}
        />
      </AppContext.Provider>,
    );

    const PropertyBookNowElement = screen.getByTestId('PropertyBookNow');
    expect(PropertyBookNowElement).toBeInTheDocument();
  });

  test('Should fill all form', () => {
    const { container } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyBookNow
          property={{
            id: '1',
            name: 'Name 1',
            description: 'Description 1',
            city: 'City 1',
            country: 'Country 1',
            purpose: 'Purpose 1',
            price: 1,
            images: ['image'],
            position: {
              latitude: 0,
              longitude: 0,
            },
          }}
        />
      </AppContext.Provider>,
    );
    const InputNameElement = screen.getByPlaceholderText('Name');
    act(() => {
      fireEvent.change(InputNameElement, { target: { value: 'Name' } });
    });
    const InputEmailElement = screen.getByPlaceholderText('Email');
    act(() => {
      fireEvent.change(InputEmailElement, { target: { value: 'email' } });
    });

    const InputDateRangeElement = screen.getByPlaceholderText('Select a date to book');
    act(() => {
      InputDateRangeElement.focus();
    });
    act(() => {
      fireEvent.click(container.getElementsByClassName('react-datepicker__day--today')[0]);
    });
    act(() => {});

    const Button = screen.getByText('Book now');
    act(() => {
      Button.click();
    });
  });
});
