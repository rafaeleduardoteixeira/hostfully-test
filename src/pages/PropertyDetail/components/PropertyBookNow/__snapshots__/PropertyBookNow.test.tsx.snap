// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyBookNow Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="sc-braxZu gEhGNz"
    data-testid="PropertyBookNow"
  >
    <div
      class="sc-gJhJTp gZBfkf"
    >
      $1
    </div>
    <div
      class="sc-fQpRED cNieMC"
    >
      <div
        class="sc-la-DxNn jVoYNI"
      >
        <input
          class="sc-knesRu kNBFpF"
          data-testid="Input"
          error=""
          name="name"
          placeholder="Name"
          type="text"
          width="300px"
        />
      </div>
      <div
        class="sc-la-DxNn jVoYNI"
      >
        <input
          class="sc-knesRu kNBFpF"
          data-testid="Input"
          error=""
          name="email"
          placeholder="Email"
          type="email"
          width="300px"
        />
      </div>
      <div
        class="sc-kkmypM gUXkBX"
        error=""
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container react-datepicker__view-calendar-icon"
          >
            <img
              alt="calendar"
              class=" react-datepicker__calendar-icon "
              src="/icons/calendar.svg"
            />
            <input
              class="date-picker"
              placeholder="Select a date to book"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <button
      class="sc-elDIKY hgUBA-D"
    >
      Book now
    </button>
  </div>
</DocumentFragment>
`;
